### DISCUSSION TITLE
求助！为防止导航属性加载不完整，系统启动初始化，但占用内存太高

### DISCUSSION THREAD

**Discussion by guochen2** [dotnetcore/FreeSql#941] [Q&A]
- 预热实体  内存1.3G+
```
  App.EffectiveTypes.Where(a => a.Namespace != null && a.Namespace.Equals("***.Entity"))
    .ToList().ForEach(a =>
    {
        fsql.CodeFirst.GetTableByEntity(a);
    });
```
- 查询预热 内存900M+
```
fsql.GetRepository<***.UserInfo>().Select.Any();
```

**Comment by 2881099** [comment]
这个跟机器本身内存有关，越多占用就越大。

App.EffectiveTypes 这个特别占内存的

**Reply by guochen2** [reply]
> 这个跟机器本身内存有关，越多占用就越大。
> 
> App.EffectiveTypes 这个特别占内存的

那目前我使用的这两种方式，对 防止导航属性加载不完整 这个问题是否都可以解决呢,还是说可以使用其他的方式？

**Reply by 2881099** [reply]
别担心，1.3G 很普通啦，如果换一台运行的机器只有2G内存，它就不会占这么多啦。跟运行的机器有关

**Reply by guochen2** [reply]
> 别担心，1.3G 很普通啦，如果换一台运行的机器只有2G内存，它就不会占这么多啦。跟运行的机器有关

要是限制docker容器的内存在1.2G  程序都跑不起来，在初始化的时候直接挂了。。

**Reply by 2881099** [reply]
你试试挂载一个未使用 freesql 的 app 到 docker，看下占用多少内存（记得要包含你已存在的所有 dll 和相关 class）

```c#
App.EffectiveTypes.Where(a => a.Namespace != null && a.Namespace.Equals("***.Entity"))
    .ToList().ForEach(a =>
    {
        Console.WriteLine(a.FullName);
    });
```

**Reply by guochen2** [reply]
@2881099  启动阶段如果不调用Freesql，内存占用130M，启动阶段调用一下Freesql的任意表的select，内存飙升到1.2G+，应该是缓存导航属性这些信息占用了大量内存，如果Entity比较少，内存也不会很高（其他项目10个entity的，内存200M左右，导航属性也不多），我们有54个Entity，我用dotMemory查看导出的dump文件，没有看到是哪个大对象占用内存。

**Reply by 2881099** [reply]
有没有开启延迟加载功能，以及反射批量注入

**Reply by guochen2** [reply]
> 有没有开启延迟加载功能，以及反射批量注入

有开启延迟加载，没有反射批量注入

**Reply by 2881099** [reply]
延迟加载是很占内存的

**Reply by guochen2** [reply]
> 延迟加载是很占内存的

好的，我尝试关闭延迟加载，测试看还占用多少资源

**Reply by guochen2** [reply]
> 延迟加载是很占内存的

移除延迟加载后启动阶段内存保持在200M，目前情况已缓解
